### DISCUSSION TITLE
python: usage with apis expecting pydantic types

### DISCUSSION THREAD

**Discussion by joprice** [fable-compiler/Fable#3970] [General]
Is there any pattern for generating a type that is annotated in a way that is usable by pydantic. Some libraries rely on it for generating a schema by calling `model_json_schema` on the type. Pydantic defines its own dataclasses, but the docs say that it has some compatibility with stdlib dataclasses:

https://docs.pydantic.dev/latest/concepts/dataclasses/#usage-of-stdlib-dataclasses-with-basemodel

I've tried a few combinations to generate a class that it likes, but haven't yet been able to get it to recognize its properties.

**Comment by joprice** [comment]
Related discussion here, since being able to add arbitrary annotations might solve one piece of this https://github.com/fable-compiler/Fable.Python/issues/48

**Comment by joprice** [comment]
Since it would be difficult to get things like discriminated unions and string enums working with pedantic, I'm going a different route for now of generating the json schema using a myriad plugin.

**Comment by dbrattli** [comment]
I think this is a great idea that we should explore. I don't have the capacity right now, but I'll keep it in mind.

**Comment by dbrattli** [comment]
@joprice This should now be fixed with #4199. E.g:

```fs
[<Erase>]
type Field<'T> = 'T

[<Import("Field", "pydantic")>]
let Field (description: string): Field<'T> = nativeOnly

[<Import("BaseModel", "pydantic")>]
type BaseModel () = class end

// Test PythonClass attribute with attributes style
[<Py.ClassAttributes(style="attributes", init=false)>]
type AttributesUser() =
    inherit BaseModel()
    member val Name: Field<string> = Field("Name") with get, set
    member val Age: int = 10 with get, set
    member val Email: string option = None with get, set
```

Now becomes:

```python
class AttributesUser(BaseModel):
    Age: int32 = int32(10)
    Email: str | None = None
    Name: str = Field("Name")
```

Still need some work on the `Field` handling to get the parameter name there. Looking into it ...

**Comment by dbrattli** [comment]
This next PR is also relevant to add Python decorators to classes: https://github.com/fable-compiler/Fable/pull/4200

**Comment by dbrattli** [comment]
Some more improvements here: https://github.com/fable-compiler/Fable/pull/4207

**Comment by dbrattli** [comment]
Binding now available in Fable.Python: https://github.com/fable-compiler/Fable.Python/blob/main/src/pydantic/Pydantic.fs
